### PR TITLE
system-variables: add tidb_enable_paging

### PR DESCRIPTION
**Pull request by you06**
Signed-off-by: you06 <you1474600@gmail.com>

### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

Add a new environment variable which controls coprocessor paging for `IndexLookUp` executor, ref pingcap/tidb#30578.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: pingcap/docs#7352
- Other reference link(s): 

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: system-variables.md [hunk 1]
```diff
@@ -1522,3 +1522,9 @@ set tidb_slow_log_threshold = 200;
 - 作用域：SESSION | GLOBAL
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tidb_enable_paging`
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求。
```
Comment (youjiali1995) [inline system-variables.md:1530]: 介绍一下 paging 呗，好处是啥，适用什么场景

**Review by youjiali1995 — APPROVED** [review]

**Comment by you06** [comment]
@TomShawn PTAL.

File: system-variables.md [hunk 2]
```diff
@@ -1522,3 +1522,12 @@ set tidb_slow_log_threshold = 200;
 - 作用域：SESSION | GLOBAL
 - 默认值：`ON`
 - 这个变量用于控制计算窗口函数时是否采用高精度模式。
+
+### `tidb_enable_paging`
```
Comment (ran-huang) [inline system-variables.md:1526]:
```suggestion
### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
```

Comment (ran-huang) [inline system-variables.md:1526]: Could you please move this sysvar to L641, just above `tidb_enable_parallel_apply`? 
It'll be easier for users to locate it if we sort all sysvars in alphabetical order.

File: system-variables.md [hunk 3]
```diff
@@ -639,6 +639,16 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求。
+- 开启 `tidb_enable_paging` 能够优化使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求。
+- 在使用无法下推 `Limit` 的 `IndexLookUp` 算子的回表场景下，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。
+- 当 `Limit` 算子的限制只需要少部分数据时，开启 `tidb_enable_paging`，能够减少处理数据的数量，实现降低延迟，降低资源消耗的效果。
+- 开启 `tidb_enable_paging` 之后，`IndexScan` 算子分 batch 返回结果，能够提前开始执行 `TableScan`，降低延迟，减少 `IndexScan` 和 `TableScan` 扫描的数据量。
```
Comment (ran-huang) [inline system-variables.md:650]: L649 和 L650 是不是重复了？

**Comment by you06** [comment]
@youjiali1995 PTAL again.

File: system-variables.md [hunk 4]
```diff
@@ -639,6 +639,15 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
 > 该变量只有在默认值 `OFF` 时，才算是安全的。因为设置 `tidb_enable_noop_functions=1` 后，TiDB 会自动忽略某些语法而不报错，这可能会导致应用程序出现异常行为。例如，允许使用语法 `START TRANSACTION READ ONLY` 时，事务仍会处于读写模式。
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求。
+- 开启 `tidb_enable_paging` 能够优化使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求。
+- 在使用无法下推 `Limit` 的 `IndexLookUp` 算子的回表场景下，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。
+- 当由于 `Limit` 算子的限制只需要少部分数据时，开启 `tidb_enable_paging`，能够减少处理数据的数量，实现降低延迟，降低资源消耗的效果。
```
Comment (ran-huang) [inline system-variables.md:649]:
```suggestion
- 适用场景：对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，在回表查询时，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。
```
How about this?

Comment (youjiali1995) [inline system-variables.md:649]: 是不是要缩进一层？
```suggestion
- 开启 `tidb_enable_paging` 能够优化使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求。
    - 在使用无法下推 `Limit` 的 `IndexLookUp` 算子的回表场景下，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。
    - 当由于 `Limit` 算子的限制只需要少部分数据时，开启 `tidb_enable_paging`，能够减少处理数据的数量，实现降低延迟，降低资源消耗的效果。
```

Comment (you06) [reply system-variables.md]: 改成一段话了，不手动换行。

**Review by youjiali1995 — APPROVED** [review]
LGTM

**Review by ran-huang — APPROVED** [review]

File: system-variables.md [hunk 5]
```diff
@@ -639,6 +639,13 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 >
 > 该变量只有在默认值 `OFF` 时，才算是安全的。因为设置 `tidb_enable_noop_functions=1` 后，TiDB 会自动忽略某些语法而不报错，这可能会导致应用程序出现异常行为。例如，允许使用语法 `START TRANSACTION READ ONLY` 时，事务仍会处于读写模式。
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求。
+- 适用场景：对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，在回表查询时，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。
```
Comment (cfzjywxk) [inline system-variables.md:647]: 具体 limit 多少数据建议开这个，这个要不要提一下？

**Review by cfzjywxk — APPROVED** [review]

File: system-variables.md [hunk 6]
```diff
@@ -639,6 +639,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
...
```
Comment (youjiali1995) [inline system-variables.md:647]: > 在回表查询时

和这个没关系吧

Comment (you06) [reply system-variables.md:647]: 额，只是想描述在需要回表的场景下，好像容易误解

File: system-variables.md [hunk 7]
```diff
@@ -639,6 +639,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 >
 > 该变量只有在默认值 `OFF` 时，才算是安全的。因为设置 `tidb_enable_noop_functions=1` 后，TiDB 会自动忽略某些语法而不报错，这可能会导致应用程序出现异常行为。例如，允许使用语法 `START TRANSACTION READ ONLY` 时，事务仍会处于读写模式。
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用 paging 方式发送 coprocessor 请求。
```
Comment (en-jin19) [inline system-variables.md:646]:
```suggestion
- 这个变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求。
```

File: system-variables.md [hunk 8]
```diff
@@ -639,6 +639,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 >
 > 该变量只有在默认值 `OFF` 时，才算是安全的。因为设置 `tidb_enable_noop_functions=1` 后，TiDB 会自动忽略某些语法而不报错，这可能会导致应用程序出现异常行为。例如，允许使用语法 `START TRANSACTION READ ONLY` 时，事务仍会处于读写模式。
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求。
+- 适用场景：对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能出现延迟高、TiKV 的 unified read pool CPU 使用高的情况。在这种情况下，由于 `Limit` 算子限制只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，降低延迟，减少资源消耗。
+    - 开启 `tidb_enable_paging` 后，`Limit` 无法下推且数量小于 960 的 `IndexLookUp` 请求会使用 paging 方式发送 Coprocessor 请求，`Limit` 的值越小优化效果越明显。
```
Comment (en-jin19) [inline system-variables.md:648]:
```suggestion
- 适用场景：对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。
    - 开启 `tidb_enable_paging` 后，`Limit` 无法下推且数量小于 `960` 的 `IndexLookUp` 请求会使用 paging 方式发送 Coprocessor 请求。`Limit` 的值越小，优化效果会越明显。
```

**Review by TomShawn — APPROVED** [review]

File: system-variables.md [hunk 9]
```diff
@@ -639,6 +639,14 @@ MPP 是 TiFlash 引擎提供的分布式计算框架，允许节点之间的数
 >
 > 该变量只有在默认值 `OFF` 时，才算是安全的。因为设置 `tidb_enable_noop_functions=1` 后，TiDB 会自动忽略某些语法而不报错，这可能会导致应用程序出现异常行为。例如，允许使用语法 `START TRANSACTION READ ONLY` 时，事务仍会处于读写模式。
 
+### `tidb_enable_paging` <span class="version-mark">从 v5.4.0 版本开始引入</span>
+
+- 作用域：SESSION | GLOBAL
+- 默认值：`OFF`
+- 这个变量用于控制 `IndexLookUp` 算子是否使用分页 (paging) 方式发送 Coprocessor 请求。
+- 适用场景：对于使用 `IndexLookUp` 和 `Limit` 并且 `Limit` 无法下推到 `IndexScan` 上的读请求，可能会出现读请求的延迟高、TiKV 的 Unified read pool CPU 使用率高的情况。在这种情况下，由于 `Limit` 算子只需要少部分数据，开启 `tidb_enable_paging`，能够减少处理数据的数量，从而降低延迟、减少资源消耗。
+    - 开启 `tidb_enable_paging` 后，`Limit` 无法下推且数量小于 `960` 的 `IndexLookUp` 请求会使用 paging 方式发送 Coprocessor 请求。`Limit` 的值越小，优化效果会越明显。
```
Comment (TomShawn) [inline system-variables.md:648]:
```suggestion
- 开启 `tidb_enable_paging` 后，`Limit` 无法下推且数量小于 `960` 的 `IndexLookUp` 请求会使用 paging 方式发送 Coprocessor 请求。`Limit` 的值越小，优化效果会越明显。
```

**Pull request merged**